### PR TITLE
[Core] Update RequestFailedException to use source generation for serialization

### PR DESCRIPTION
**Pull request by m-redding**
The changes here update `RequestFailedException.TryExtractErrorContent` to use source generation for serialization instead of reflection-based serialization. It takes advantage of the fact that the [source generation is included in the net6.0+ runtime](https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/overview#how-to-get-the-library) so we don't have to upgrade our dependency on `System.Text.Json` to take advantage of this feature (😄 ).

For more context on source generation see the following:
- [How to use source generation](https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/source-generation)
- [Source generation vs reflection-based serialization](https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/source-generation-modes?pivots=dotnet-8-0)
- [Introduction of source generation](https://devblogs.microsoft.com/dotnet/try-the-new-system-text-json-source-generator/)

**Important note** - source generation does not support private classes or members, so a few classes and members needed for serialization had to be changed from private to internal. See [known issues](https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/source-generation-modes?pivots=dotnet-8-0#known-issues)

This is the last change on the `Azure.Core` side needed to unlock Native AOT with the following libraries:
- Open Telemetry Exporter - https://github.com/Azure/azure-sdk-for-net/issues/35572 / https://github.com/Azure/azure-sdk-for-net/issues/37761
- KeyVault - https://github.com/Azure/azure-sdk-for-net/issues/38500
- Storage

And likely many other packages across our SDK's.

# Contributing to the Azure SDK

Please see our [CONTRIBUTING.md](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md) if you are not familiar with contributing to this repository or have questions.

For specific information about pull request etiquette and best practices, see [this section](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#pull-request-etiquette-and-best-practices).

### CODE REVIEW

**Comment by m-redding** [comment]
cc: @vitek-karas @TimothyMothra @eerhardt

**Review by TimothyMothra — APPROVED** [review]
LGTM.
This is the same change I'm making in the Exporter. :)

Edit: reset my approval since I'm not a maintainer of Azure.Core

**Review by TimothyMothra — COMMENTED** [review]
LGTM.
This is the same change I'm making in the Exporter. :)

Edit: reset my approval since I'm not a maintainer of Azure.Core

File: sdk/core/Azure.Core/src/ResponseError.cs [hunk 1]
```diff
@@ -18,7 +18,8 @@ namespace Azure
     [TypeReferenceType(true, new string[] { nameof(Target), nameof(Details) })]
     public sealed class ResponseError
     {
-        private readonly JsonElement _element;
+        // This field needs to be internal rather than private so that it can be used by the System.Text.Json source generator
+        internal readonly JsonElement _element;
```
Comment (eerhardt) [inline sdk/core/Azure.Core/src/ResponseError.cs:22]: This is surprising to me. Doesn't the source generator pick up the `SerializationConstructor`?

File: sdk/core/Azure.Core/src/ResponseError.cs [hunk 2]
```diff
@@ -84,7 +86,7 @@ private class Converter : JsonConverter<ResponseError?>
                 return Read(element);
             }
 
-            private static ResponseError? Read(JsonElement element)
+            internal static ResponseError? Read(JsonElement element)
```
Comment (eerhardt) [inline sdk/core/Azure.Core/src/ResponseError.cs:89]: Why does this need to be `internal`?

File: sdk/core/Azure.Core/src/RequestFailedException.cs [hunk 3]
```diff
@@ -259,8 +259,26 @@ internal static bool TryExtractErrorContent(Response response, out ResponseError
...
+                    error = errorResponse?.Error;
+                }
+                if (error == null)
+                {
+                    var deserializedResponseError = System.Text.Json.JsonSerializer.Deserialize(content, typeof(ResponseError), ResponseErrorSourceGenerationContext.Default);
+                    if (deserializedResponseError != null)
+                    {
+                        error = (ResponseError)deserializedResponseError;
+                    }
+                }
```
Comment (eerhardt) [inline sdk/core/Azure.Core/src/RequestFailedException.cs:277]: Why can't this code be written like the code in the `#else` section?

```suggestion
                error = System.Text.Json.JsonSerializer.Deserialize<ErrorResponse>(content, ResponseErrorSourceGenerationContext.Default.ErrorResponse)?.Error;
                error ?? = System.Text.Json.JsonSerializer.Deserialize<ResponseError>(content, ResponseErrorSourceGenerationContext.Default.ResponseError);
```

Comment (m-redding) [reply sdk/core/Azure.Core/src/RequestFailedException.cs:277]: I can try it again, it didn't look like the generic version has an overload that takes in a JsonSerializerContext in .NET 6 - https://learn.microsoft.com/en-us/dotnet/api/system.text.json.jsonserializer.deserialize?view=net-6.0 so it wasn't building

Comment (eerhardt) [reply sdk/core/Azure.Core/src/RequestFailedException.cs:277]: This should be the one you will want to use:

https://learn.microsoft.com/en-us/dotnet/api/system.text.json.jsonserializer.deserialize?view=net-6.0#system-text-json-jsonserializer-deserialize-1(system-readonlyspan((system-char))-system-text-json-serialization-metadata-jsontypeinfo((-0)))

```
public static TValue? Deserialize<TValue> (ReadOnlySpan<char> json, System.Text.Json.Serialization.Metadata.JsonTypeInfo<TValue> jsonTypeInfo);
```

Comment (m-redding) [reply sdk/core/Azure.Core/src/RequestFailedException.cs]: Ahh ok I see, thanks!

File: sdk/core/Azure.Core/src/ResponseError.cs (hunk 1 above)
Comment (m-redding) [reply sdk/core/Azure.Core/src/ResponseError.cs:22]: I'll try without - I had been getting this vague error - `SYSLIB1030: System.Text.Json source generator did not generate output for type ResponseError` and it was a bit of trial and error trying to figure out why.

**Review by eerhardt — APPROVED** [review]
Assuming this works and passes all tests, LGTM.

**Review by JoshLove-msft — APPROVED** [review]

**Review by vitek-karas — APPROVED** [review]
Nice

**Review by christothes — APPROVED** [review]

**Pull request merged**